feat(app): field service meetings

ild0tt0re opened the pull request:
# Description

[[FEAT] Meetings for field service](https://app.clickup.com/t/86c3xf6gb) (Part 1: UI Only) 

## Type of change

This PR introduces the UI implementation for the new **Meetings for field service** feature.
Note: This update includes only the user interface components to start grabbing feedback—no backend logic or data integration has been added yet.

All relevant screens, forms, and controls for the feature are present, except for **publish** and **edit mode**
UI elements are styled and responsive according to the design specifications.
No functional logic or API connections are included at this stage.
Further work will be required to connect the UI to the backend and implement the feature’s business logic.


- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings

ild0tt0re committed:
feat(field_service_meeting): adds week view

ux-git commented on src/definition/field_service_meetings.ts:
```diff
@@ -0,0 +1,89 @@
+  Publisher,
+  KingdomHall,
+  Territory,
+  Zoom,
```
I agree, Online sounds short and versatile

ux-git committed:
refactor(field_service_meetings): code quality cleanup pass

- Remove docs/AGENTS.md and docs/* scaffolding (AI-generated, unrelated
  to the feature and not appropriate for an OSS PR)
- Delete duplicate vite.config.js (was shadowing canonical vite.config.ts;
  Vite resolves .js before .ts)
- Revert package-lock.json to merge-base (no package.json changes; only
  "peer": true churn from different npm versions)
- Fix TS2322: add ariaLabel?: string to ButtonPropsType and forward it as
  aria-label on MUIButton (resolves the one new tsc error)
- Remove duplicate ...upcomingEventsSchema spread in appDb.ts v13 block
- Delete orphan i18n key tr_fieldServiceMeetingCategory_group (zero refs)
- Extract getGroupBadgeColor() shared util (deduplicate sort_index % 10
  formula from meeting_item and useMonthView)
- Extract filterMeetingsByDataView() shared util (deduplicate identical
  dataView filter from useFieldServiceMeetings and useScheduleExport)
- Drop as any cast in locationIcons.tsx via ReactElement<{color?}> typing
- Name magic number 12 as MAX_WEEKS_LOOKAHEAD in recurringPrefill.ts

ux-git committed:
fix(post-rebase): remove duplicate declarations and add missing prop

- Remove duplicate ReactElement import and NavBarOptionsType in
  src/definition/app.ts (both sides independently added them during
  rebase of the new navbar feature)
- Remove duplicate navBarOptionsState atom in src/states/app.ts for
  the same reason
- Remove duplicate aria-label={props.ariaLabel} in Button component
  (local ariaLabel const already holds the same value)
- Add required lang={lang} prop to <Document> in FSM PDF template

tsc error count: back to the origin/main baseline (414 errors, all
pre-existing @emotion/styled and missing-package issues unrelated to
this branch).

ux-git committed:
feat(field_service_meetings): full feature overhaul — UX, a11y, and cleanup

Complete implementation of the Field Service Meetings feature including:

**Data & business logic**
- Conductor stored as person_uid with display-name resolution fallback
- Service overseer auto-resolved from group data
- Backup/sync wired; schedule Publish removed (auto-sync only)
- Recurring meeting times per-group (quick settings + settings page)
- Editing rights via usePermissions (group overseer/admin/service-overseer)
- Similar-meeting duplicate warning dialog before save
- ICS add-to-calendar download

**UI / layout**
- New header: NavBarButton pattern, blue Add button, bottom-nav padding
- Week + month calendar views with group-coloured badges
- Month view: keyboard-accessible day cells (role, tabIndex, onKeyDown)
- Calendar nav: week/month toggle full-width on mobile, below date nav
- Meeting item: always-horizontal date+info row; full-width add-to-calendar on mobile; smaller address text on mobile; real group badge colours
- Form: 3-col laptop / 2-col mobile responsive layout; inline conductor validation; delete confirmation dialog; Cancel blue in add mode; role=form a11y landmark

**Rename & i18n**
- "Regular field service meeting" → "Field service meeting"
- Group badge label: name-only or "Group N" (no "Group N – Name" compound)
- New locale keys: tr_deleteMeeting, tr_deleteMeetingDesc, tr_previousPeriod, tr_nextPeriod, tr_openCurrentWeek, tr_openCurrentMonth

**Code cleanup**
- Deleted dead page_header and schedule_publish components
- Removed canManageMeetings, getFilteredMeetings, weekRangeLabel (from return), stale index types
- IconDate: removed nested SVG + static mask ID (fixed invisible icon bug)
- ButtonGroup: fullWidth prop support

ux-git committed:
fix(time_picker): show current time when no value; fix arrow colors in dark theme

- Add referenceDate={new Date()} to DesktopTimePicker so the clock popup
  shows the current time as the starting point when value is null instead
  of displaying an empty/midnight state
- Change .MuiPickersArrowSwitcher-button color from var(--accent-150) to
  var(--accent-350): accent-150 was near-white in light mode and near-black
  in dark mode (both invisible); accent-350 is a readable medium-grey in
  both themes
- Recover FSM PDF template types (index.types.ts) and views/index.ts export
  (TemplateFieldServiceMeetings) that were dropped when the rebase squashed
  the original de0ddbf7 commit as empty

ux-git committed:
fix(sonar): address 21 of 22 code quality issues from CI review

- Remove redundant 'as GroupOption' cast in useMeetingForm (S4325)
- Replace .sort() chains with .toSorted() in useScheduleExport (S4043)
- Remove redundant FieldServiceMeetingType[] cast in backupUtils (S4325)
- Simplify arrow to Number reference in recurringPrefill (S7770)
- Use == null instead of === null || === undefined (S6582)
- Use nullish coalescing ?? in useRecurringTimes (S6606)
- Flip negated ternary conditions to positive form (S7735) in
  week_range_selector, FSM page paddingBottom, month_view aria-label
- Use replaceAll() and String.raw in icalendar.ts (S7781, S7780)
- Replace removeChild() with link.remove() in icalendar.ts (S7762)
- Use stable dateStr key instead of array index in month_view (S6479)

Skipped: S3776 cognitive complexity on month_view render (requires
major refactor — tracked separately)

ux-git commented on src/features/congregation/field_service_meetings/meeting_form/index.tsx:
```diff
@@ -0,0 +1,428 @@
+  FieldServiceMeetingCategory,
+  FieldServiceMeetingType,
+  FIELD_SERVICE_MEETING_CATEGORY_TRANSLATION_KEYS,
+} from '@definition/field_service_meetings';
```
Import retained — `FieldServiceMeetingCategory` is used as a generic type parameter on `<Autocomplete<FieldServiceMeetingCategory>>` at line 145. Removing it breaks the build (TS2304). The 21 other Sonar issues have been addressed in the latest commit.

ux-git committed:
fix(sonar): address remaining code quality issues

- Use import type for FieldServiceMeetingCategory (type-only usage
  as JSX generic parameter — satisfies linter without breaking build)
- Restructure Map get()+! pattern to get()+if(!entry){set} so
  TypeScript narrows without non-null assertions (S4325 x2)
- Use optional chain for recurring null guard in recurringPrefill
  (S6582): recurring?.weekday.value == null handles both null and
  undefined recurring in one expression
- Extract nested ternary to meetingSuffix variable in month_view
  aria-label (S3358)

Remaining: S3776 cognitive complexity on month_view render (44 vs
allowed 15) — requires major refactor, tracked separately

ux-git commented on src/features/congregation/field_service_meetings/meeting_form/form_fields.tsx:
```diff
@@ -0,0 +1,428 @@
+  FieldServiceMeetingCategory,
+  FieldServiceMeetingType,
+  FIELD_SERVICE_MEETING_CATEGORY_TRANSLATION_KEYS,
+} from '@definition/field_service_meetings';
```
Fixed: changed to `import type { FieldServiceMeetingCategory }` — it's used only as a JSX generic type parameter, so a type-only import satisfies the linter while keeping the type reference intact.

ux-git committed:
refactor(month_view): reduce cognitive complexity (S3776)

Extract DayCell sub-component and getCellBorderSx helper to bring
the main render function's cognitive complexity from 41 down to the
allowed 15. Behaviour is unchanged — same conditional logic, same
border rules, same interaction handling — just split across smaller
focused units.

ux-git committed:
refactor(month_view): further reduce DayCell cognitive complexity (S3776)

- Extract MeetingBadges sub-component (removes tabletUp branch and
  hidden>0 conditional from DayCell)
- Move handleDayCellKeyDown to module scope (removes nesting penalty)
- Use 'cell !== undefined' narrowing instead of Boolean(cell)+! so
  TypeScript narrows naturally — removes unnecessary assertion (S4325)

ux-git committed:
refactor(month-view): pre-compute conditional props to cut cognitive complexity

Pre-compute all conditional values (role, tabIndex, onClick, onKeyDown,
ariaLabel, dotSize, cellSx) above the JSX return in DayCell so every
ternary sits at nesting depth 0 (cost 1 each) instead of inside the
JSX tree where MUI sx nesting would multiply the penalty.

Target: ≤15 (Sonar S3776), previously reported at 31.

ux-git committed:
refactor(month-view): extract getDayCellConfig helper to eliminate DayCell branching

Move all conditional logic (layout, interactive props, sx computation)
into a module-level getDayCellConfig() function so DayCell itself
contains only two &&-checks (isToday, hasMeetings).

DayCell cognitive complexity: 28 → 2 (target ≤15, S3776).
getDayCellConfig complexity: ~8 (also within limit).

ux-git committed:
refactor(field-service-meetings): senior-review fixes and consistency pass

Correctness/data fixes:
- stop mutating shared week-range atom via getWeekDate (pass copies)
- align week list/export to full meetings state so past periods show
- fix month-navigation day overflow; share editing id via atom
- stop writing legacy top-level _deleted/updatedAt; guard cleanup migration
- broaden FSM sync export gate to match edit permissions
- wire FSM into local file import/export
- guard single-day joint-meeting date range; drop referenceDate on TimePicker

Reuse/consistency:
- reuse @components/card; dedupe group-badge colour + persist helpers
- replace hand-rolled icalendar util with the ics package
- normalise feature to house structure (snake_case folders, index.types.ts,
  container/hook split); remove banner/vacuous comments and FC typing
- delete dead IconEventAvailable; fix icon generator round-trip
- remove duplicate/unused locale keys

ux-git committed:
refactor(field-service-meetings): address SonarQube findings

- time_picker: extract withBaseSx helper to module level so the sx-merge
  branch no longer counts toward the component's cognitive complexity
  (16 -> 15, back within the allowed threshold)
- useFieldServiceMeetingsImport: use .some() instead of .find() for the
  presence check (the result was only used as a boolean)

ux-git committed:
refactor(meeting-form): infer Autocomplete generics instead of passing them

The three Autocomplete usages passed explicit type arguments
(<Autocomplete<GroupOption>>, etc.), unlike all 14 other consumers in the
codebase which rely on inference from `options`/`value`. The explicit
generics were redundant — the option type infers identically, including the
freeSolo conductor field — and forced type-only imports that static analysis
kept flagging as unused. Removing them matches house style and drops the
imports.

ux-git committed:
Merge remote-tracking branch 'origin/main' into feat/field-service-meetings

# Conflicts:
#	src/definition/app.ts

ux-git committed:
fix(field-service-meetings): rebuild the PDF export to match the design

The template hand-rolled its own Page, header and font registration instead of
the shared PDF components, and its table was a set of ad-hoc percentages and
hex values that matched neither the Figma frame nor the app tokens.

It now renders through Document/Page/PageContent/PageHeader like the other
exports, and the table follows the design grid: a 129pt month tab, a 56/64pt
date and time column, two 207.5pt text columns, accent-200 rules and an
accent-150 date cell.

The header icon was the ministry symbol; the design uses the service one, so a
react-pdf IconService joins the shared PDF icon set. The right-hand pill now
carries the group label the export is scoped to, and date cells drop the
redundant month name that the month tab already states.

Rows and month headers no longer split across pages, and the columns are sized
in points, so long addresses wrap on word boundaries instead of collapsing to
one word per line.

ux-git committed:
style(field-service-meetings): left-align the date and time columns in the PDF

Centred date and time cells made the eye jump between column starts when
scanning a month. Both now start on the left, on the same 4pt inset the
address and conductor cells use, and their headers move with them.

ux-git committed:
style(field-service-meetings): inset the PDF date column by a further 4px

The date column sits against the table's left border, so its 4px inset read
tighter than the other columns, which have a divider's worth of space in front
of them. Header and cells now start 8px in.

ux-git commented on src/components/group_badge/index.tsx:
```diff
@@ -0,0 +1,45 @@
+        sx={align === 'center' ? { textAlign: 'center' } : undefined}
+      >
+        {label}
+      </Typography>
```
Done in 94457fd — `minWidth: 0` on the Typography so the label truncates with `fullWidth`.

ux-git commented on src/components/icons/IconDate.tsx:
```diff
@@ -16,35 +16,29 @@ const IconDate = ({
       className={`organized-icon-date ${className}`}
+      viewBox="0 0 24 24"
       sx={{ width: `${width}px`, height: `${height}px`, ...sx }}
     >
```
Done in 94457fd — `className ?? ''` plus a trim, so no literal `undefined` in the class list.

ux-git commented on src/components/time_picker/slots/textfield.tsx:
```diff
@@ -9,6 +9,14 @@ const InputTextField = forwardRef(function DatePickerInputField(
+      ? (props.sx as { height?: string }).height
+      : undefined;
+  const customHeight = consumerHeight ?? `${heightLocal}px`;
+
```
Done in 94457fd — the padding and label offset are now derived from the height the input actually renders at, with a fallback to 44px for non-px values.

ux-git commented on src/features/congregation/field_service_meetings/index.tsx:
```diff
@@ -0,0 +1,122 @@
+        onClose={handleCancelEdit}
+        onDelete={handleDeleteMeeting}
+      />
+    ) : undefined;
```
Skipped — the add form is already rendered by the page (`src/pages/congregation/field_service_meetings/index.tsx`) when `isCreating` is true. The container only owns the edit form, which is why it clears the editing uid.

ux-git commented on src/features/congregation/field_service_meetings/meeting_form/index.tsx:
```diff
@@ -0,0 +1,410 @@
+    setDeleteConfirmOpen(false);
+    if (!onDelete) return;
+    await onDelete(meeting);
+  };
```
Done in 94457fd — the delete button renders only when `onDelete` is passed and is disabled while the deletion is in flight. The footer was restructured so edit mode no longer risks two cancel buttons.

ux-git commented on src/features/congregation/field_service_meetings/meeting_form/useMeetingForm.tsx:
```diff
@@ -0,0 +1,413 @@
+    formData.meeting_data.group_id,
+    formData.meeting_data.type,
+    groupOptions,
+  ]);
```
Done in 94457fd — a meeting's own group stays in the options even after the group is deleted, so editing no longer silently turns it into a joint meeting.

ux-git commented on src/features/congregation/field_service_meetings/meeting_form/useMeetingForm.tsx:
```diff
@@ -0,0 +1,413 @@
+      });
+    },
+    [formData.meeting_data.start, meetingDuration, updateMeetingData]
+  );
```
Done in 94457fd — an unparsable stored start falls back to now, and the update bails out if the date is still invalid after the change.

ux-git commented on src/features/congregation/field_service_meetings/meeting_item/index.tsx:
```diff
@@ -0,0 +1,309 @@
+        gap: '12px',
+        padding: tabletUp ? '24px' : '16px',
+      }}
+    >
```
Done in 94457fd, with a follow-up in 28891e5: both controls also reveal on `:focus-within`, and the hidden-until-hover state now applies only where hover exists — a desktop-width touch device kept them invisible for good.

ux-git commented on src/features/congregation/field_service_meetings/recurring_prefill.ts:
```diff
@@ -0,0 +1,76 @@
+        meeting.meeting_data.group_id === groupId &&
+        formatDate(new Date(meeting.meeting_data.start), 'yyyy/MM/dd') ===
+          dateStr
+    );
```
Done in 94457fd — the availability check falls back to `meeting_data.type`, matching how the form resolves the selected group.

ux-git commented on src/features/congregation/field_service_meetings/schedule_export/useScheduleExport.tsx:
```diff
@@ -0,0 +1,286 @@
+          sortKey: formatted.start,
+          days: new Map(),
+        };
+        monthMap.set(key, monthEntry);
```
Done in 94457fd — the bucket key and the month title both use the year of the date being placed.

ux-git commented on src/features/congregation/field_service_meetings/usePermissions.tsx:
```diff
@@ -0,0 +1,110 @@
+      }
+    }
+    return ids;
+  }, [groups, userUID]);
```
Skipped — `fieldWithLanguageGroupsState` already filters `group_data._deleted`, so a deleted group never reaches this hook and cannot confer leadership.

ux-git commented on src/features/congregation/field_service_meetings/usePermissions.tsx:
```diff
@@ -0,0 +1,110 @@
+        return false;
+      }
+
+      return isAnyGroupLead;
```
Done in 94457fd — the branch is intended (a group lead may edit a visit scheduled for his own group; creating one stays with the service overseer and admins via the form's category options), so the header doc now says so and the behaviour is unchanged.

ux-git commented on src/features/congregation/settings/field_service_meeting_settings/recurring_times/useRecurringTimes.tsx:
```diff
@@ -0,0 +1,93 @@
+
+    await dbAppSettingsUpdate({
+      'cong_settings.field_service_meeting_times': arr,
+    });
```
Done in a631c8d — the read-modify-write moved into `dbAppSettingsUpdateFieldServiceMeetingTime`, a Dexie transaction that reads the stored settings instead of the render snapshot.

ux-git commented on src/components/time_picker/slots/textfield.tsx:
```diff
@@ -7,7 +7,25 @@ const InputTextField = forwardRef(function DatePickerInputField(
+  // height for non-px units, which can't be resolved here.
+  const parsedHeight = Number.parseFloat(customHeight);
+  const effectiveHeight =
+    customHeight.trim().endsWith('px') && Number.isFinite(parsedHeight)
```
Done in 59a03ee — the slot resolves a pixel height where it can (string px or a number above 1) and falls back to the local 44px for percentages, other units and responsive objects, so no string method runs on a non-string.

ux-git commented on src/services/dexie/metadata.ts:
```diff
@@ -27,6 +27,7 @@ export const dbMetadataDefault = async () => {
           user_field_service_reports: { version: '', send_local: true },
           upcoming_events: { version: '', send_local: true },
           delegated_field_service_reports: { version: '', send_local: true },
+          field_service_meetings: { version: '', send_local: true },
```
Done in 59a03ee — `dbMetadataDefault` now merges missing keys into an existing metadata record, so an installation predating a table gets its key on the next start instead of exporting with `send_local` undefined. `dbMetadataReset` shares the same defaults. No unit test added: the repo only runs Cypress and `runUpdater` has no harness to hook into.
